### ISSUE TITLE
Custom event listener for the blitzInject JS function

### ISSUE THREAD

**Issue by pierrestoffe** [putyourlightson/craft-blitz#179]
### Is your feature request related to a problem? Please describe.
I am installing Blitz on a website where the navigation/transition between pages is handled by XHR requests (via https://github.com/barbajs/barba). On that same website we have forms and use the `{{ craft.blitz.csrfInput() }}` method to set CSRF token dynamically.

Here is the way Blitz sets dynamic values: [a script](https://github.com/putyourlightson/craft-blitz/blob/develop/src/resources/js/blitzInjectScript.js) is injected into every single page. The function in this scripts finds all occurrences of `span.blitz-inject`, makes XHR requests to get a CSRF token for each occurence and injects the response back into the HTML. The event that initiates this function is `DOMContentLoaded`. It works perfectly on the very first page our website's visitors visit.

The issue is that when a page is loaded with XHR/barba.js the `DOMContentLoaded` event will not be triggered again. So Blitz can't generate new CSRF tokens for every pages that was loaded using a XHR request. Basically this means that any page other the page a visitor landed on initially doesn't have dynamic CSRF tokens.

The obvious solution is to call the `DOMContentLoaded` event after a page is XHR-loaded, but that won't work all the time. On this specific project `DOMContentLoaded` is listened to by other modules that should not be called again, and that actually depend on this event. I could potentially make it work of course, but I don't find it ideal.

### Describe the solution you would like
I would like to be able to define which event the Blitz script listens to. In other words: have the use Blitz's config file to replace `DOMContentLoaded` with a custom event listener.

**Comment by bencroker** [comment]
Why not simply call the `blitzInject()` function after each page load using your own event listener?

**Comment by pierrestoffe** [comment]
It is the only solution I'm using now indeed, but this creates two issues in my opinion:
- the technique will trigger errors if the `blitzInject()` function is ever renamed in Blitz
- if I am to temporarily disable the Blitz plugin, the custom event listener will be calling a function that does not exist and return an error

So with future-proofness and stability in mind I would prefer to use a safer method such as the one I described above. Happy to create a PR if you're open to this solution?

**Comment by bencroker** [comment]
I've add an `injectScriptEvent` config setting in the develop branch that you can override with a custom event. 
https://github.com/putyourlightson/craft-blitz/blob/develop/src/config.php#L184

In your use-case however, since you are loading a new page through an XHR request, will the script tags in the returned source code even be executed? I thought script tags that were injected into the DOM were ignored, or are you doing something to execute them explicitly?

**Comment by pierrestoffe** [comment]
Thanks @bencroker, it seems to work exactly as I wanted it to.

The page is loaded with XHR, but only the `<main>` tag is replaced. Blitz's script is located outside of the `<main>` tag and is never replaced. Otherwise we would probably be limited by what you described.

**Comment by bencroker** [comment]
Right but don't you want the script tag in the content loaded through the XHR request to be run? Otherwise won't the `blitzInject()` function be called on the old elements?

**Comment by pierrestoffe** [comment]
**TLDR**
There is not need to execute any script from the XHR request, no. `blitzInject` will query the new elements as soon when it is called again.

**In more details now**
When a new page is loaded with XHR:
1. We get the content of the <main> tag of the new page
2. We inject it in the <main> tag of the page that was visited (we don't touch the `<script>` that was injected by Blitz)
3. The `blitzUpdate` custom event is triggered (this is our custom event)
4. This custom event triggers the `blitzInject` function since we set `injectScriptEvent` to `blitzUpdated` in _config/blitz.php_
5. The `blitzInject` function gets elements from the **updated** DOM using `document.getElementById`
6. Blitz does its magic on these elements
_repeat on each page load..._

This is only possible because:
1. Your `<script>` queries elements (using `document.getElementById`) from inside the `blitzInject` function
2. The `blitzInject` function is re-run when the event listener is triggered

I just tested on my local environment and it works perfectly well. So the issue looks solved in my opinion.

**Comment by bencroker** [comment]
Ah very cool! Will release the update soon.

**Comment by bencroker** [comment]
Added in version 3.5.0.

**Comment by pierrestoffe** [comment]
I see what you meant now. I had not noticed that the Blitz directives were included in the <script> tag and thought they were stored in the database or in the $_SESSION. It did not cause any issue where I tested it earlier this week, but now I see the issue.

To make things work I had to update part of the mechanism used to define the Blitz directives.
See my pull request: #184

**Comment by bencroker** [comment]
Thanks for the pull request. We'll need to do some testing on this as it changes the mechanism of the Blitz inject process (and could even be considered a breaking change).

**Comment by bencroker** [comment]
Made a few tweaks to your PR in https://github.com/putyourlightson/craft-blitz/commit/0592a7715f7437533457db29098597e26be9af87, can you please let me know whether this still solves your issue?

**Comment by pierrestoffe** [comment]
Yes, it still works perfectly. No issue on my end.

**Comment by bencroker** [comment]
Great, thanks!

**Comment by bkmorse** [comment]
> **TLDR**
> There is not need to execute any script from the XHR request, no. `blitzInject` will query the new elements as soon when it is called again.
> 
> **In more details now**
> When a new page is loaded with XHR:
> 
> 1. We get the content of the  tag of the new page
> 2. We inject it in the  tag of the page that was visited (we don't touch the `<script>` that was injected by Blitz)
> 3. The `blitzUpdate` custom event is triggered (this is our custom event)
> 4. This custom event triggers the `blitzInject` function since we set `injectScriptEvent` to `blitzUpdated` in _config/blitz.php_
> 5. The `blitzInject` function gets elements from the **updated** DOM using `document.getElementById`
> 6. Blitz does its magic on these elements
>    _repeat on each page load..._
> 
> This is only possible because:
> 
> 1. Your `<script>` queries elements (using `document.getElementById`) from inside the `blitzInject` function
> 2. The `blitzInject` function is re-run when the event listener is triggered
> 
> I just tested on my local environment and it works perfectly well. So the issue looks solved in my opinion.

@pierrestoffe do you mind sharing the snippet of code that gets triggered by the event you set in the blitz.php config file for: injectScriptEvent?

**Comment by bencroker** [comment]
@bkmorse I'm guessing it is just `blitzInject();` (using the latest commit on the `develop` branch).

**Comment by pierrestoffe** [comment]
The content of config/blitz.php

```
<?php

return [
    'injectScriptEvent' => 'blitzUpdate'
];
```

The content of my JS module that call the `blitzUpdate` event

```
import { listenEvent, dispatchEvent, registerEvent } from '../helpers/events'

/**
 * @function `init`
 */
const init = () => {
    registerEvent('blitzUpdate');
    dispatchEvent('blitzUpdate', document);

    listenEvent('pageUpdate', function () {
        dispatchEvent('blitzUpdate', document);
    });
};

const module = () => {
    init();
};

export default module;
```

All the rest is handled by Blitz. The code that Blitz adds at the bottom of all pages contains:
- an event listener for the `blitzUpdate` event, which then triggers the `blitzInject ` function
- the `blitzInject` function which injects dynamic content wherever needed

**Comment by bencroker** [comment]
Added to Blitz version 3.6.1.
